### ISSUE TITLE
Integer IDs in SequenceCollection crash SequenceCollection.write()

### ISSUE THREAD

**Issue by steve-federowicz** [biocore/scikit-bio#881]
Hi,

I noticed that when I create a SequenceCollection of DNASequence() objects and use integers as the id in the DNASequence() constructor, the SequenceCollection.write() method crashes with the following.  

Maybe this is a feature and not a bug but could probably just cast every id to str or spit out a warning in the DNASequence() constructor.

-Steve

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-92-3b7b520c55b7> in <module>()
----> 1 sequences2.write('tmp.fasta')

/usr/local/lib/python2.7/dist-packages/skbio/io/_registry.pyc in write(self, fp, format, **kwargs)
    723 
    724         def write(self, fp, format=cls.default_write_format, **kwargs):
--> 725             skbio_io_write(self, into=fp, format=format, **kwargs)
    726 
    727         write.__doc__ = _write_docstring % (

/usr/local/lib/python2.7/dist-packages/skbio/io/_registry.pyc in write(obj, format, into, mode, **kwargs)
    665                                                   else str(cls)))
    666 
--> 667     writer(obj, into, mode=mode, **kwargs)
    668 
    669 

/usr/local/lib/python2.7/dist-packages/skbio/io/_registry.pyc in wrapped_writer(obj, fp, mode, **kwargs)
    338                 for key, fh in zip(file_keys, fhs[1:]):
    339                     kwargs[key] = fh
--> 340                 writer(obj, fhs[0], **kwargs)
    341 
    342         wrapped_writer.__doc__ = writer.__doc__

/usr/local/lib/python2.7/dist-packages/skbio/io/fasta.pyc in _sequence_collection_to_fasta(obj, fh, qual, id_whitespace_replacement, description_newline_replacement, max_width)
    757                                   max_width=None):
    758     _sequences_to_fasta(obj, fh, qual, id_whitespace_replacement,
--> 759                         description_newline_replacement, max_width)
    760 
    761 

/usr/local/lib/python2.7/dist-packages/skbio/io/fasta.pyc in _sequences_to_fasta(obj, fh, qual, id_whitespace_replacement, description_newline_replacement, max_width)
    831         id_whitespace_replacement=id_whitespace_replacement,
    832         description_newline_replacement=description_newline_replacement,
--> 833         max_width=max_width)

/usr/local/lib/python2.7/dist-packages/skbio/io/_registry.pyc in wrapped_writer(obj, fp, mode, **kwargs)
    338                 for key, fh in zip(file_keys, fhs[1:]):
    339                     kwargs[key] = fh
--> 340                 writer(obj, fhs[0], **kwargs)
    341 
    342         wrapped_writer.__doc__ = writer.__doc__

/usr/local/lib/python2.7/dist-packages/skbio/io/fasta.pyc in _generator_to_fasta(obj, fh, qual, id_whitespace_replacement, description_newline_replacement, max_width)
    693         obj, id_whitespace_replacement, description_newline_replacement,
    694         qual is not None)
--> 695     for header, seq_str, qual_scores in formatted_records:
    696         if max_width is not None:
    697             seq_str = _chunk_str(seq_str, max_width, '\n')

/usr/local/lib/python2.7/dist-packages/skbio/io/_base.pyc in _format_fasta_like_records(generator, id_whitespace_replacement, description_newline_replacement, require_qual)
    169         id_ = seq.id
    170         if id_whitespace_replacement is not None:
--> 171             id_ = _whitespace_regex.sub(id_whitespace_replacement, id_)
    172 
    173         desc = seq.description

TypeError: expected string or buffer
```

**Comment by ebolyen** [comment]
Hey @steve-federowicz thanks for creating an issue! This is a bit of a bug in the constructor that should be resolved relatively soon as we are going through those objects with more robust validation during #826. Our current plan is to just error when constructing a sequence with an ID that is not a `str` type, but we could be convinced to cast if you think that would be useful (as long as it was obvious such a cast was occurring).

**Comment by steve-federowicz** [comment]
Hmmmm ok well it sounds like there is a strong reason why id should be `str` and not `object` so I would say definitely just throw the error in construction.  I agree that casting blindly would not be a good idea.

Thanks!!

**Comment by ebolyen** [comment]
Your situation is basically the reason we need id's to be strings. That being said, ints have a very reasonable string interpretation that wouldn't surprise anyone (in fact the lack of an interpretation would be confusing, as evidenced by this issue). @jairideout any thoughts on this?

**Comment by jairideout** [comment]
I think IDs should be stored internally as strings but we could definitely support (some) casting of inputs for convenience. Integers definitely make sense, probably others too.

The reason why we don't want to store IDs as their native type (i.e., something that's hashable) is to prevent confusion when indexing into data structures. For example, you can currently index into `DistanceMatrix` by either ID (string) or index (integer). If IDs were stored as integers, it would be ambiguous whether lookup would be performed by ID or index. This same issue will crop up in other data structures with flexible indexing (`SequenceCollection`, `Alignment`, etc.).

Thanks for reporting this bug @steve-federowicz!

**Comment by jairideout** [comment]
This was resolved in #920. We decided to only support `str` IDs for `Sequence` objects to avoid implicit conversions that might be confusing if users are expecting the IDs to be stored/retrieved as integers. A better error message is raised now by the `Sequence` constructor, for example:

```python
n [1]: from skbio import DNA

In [2]: DNA('ACGT', id=42)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-9ded09d3551d> in <module>()
----> 1 DNA('ACGT', id=42)

/Users/jairideout/dev/scikit-bio/skbio/sequence/_iupac_sequence.pyc in __init__(self, sequence, id, description, quality, validate, case_insensitive)
    168                  validate=True, case_insensitive=False):
    169         super(IUPACSequence, self).__init__(
--> 170             sequence, id=id, description=description, quality=quality)
    171
    172         if case_insensitive:

/Users/jairideout/dev/scikit-bio/skbio/sequence/_sequence.pyc in __init__(self, sequence, id, description, quality)
    189             sequence = sequence._bytes
    190
--> 191         self._set_id(id)
    192         self._set_description(description)
    193         self._set_sequence(sequence)

/Users/jairideout/dev/scikit-bio/skbio/sequence/_sequence.pyc in _set_id(self, id_)
    198             self._id = id_
    199         else:
--> 200             raise TypeError("Sequence ID %r must be of type `str`." % (id_,))
    201
    202     def _set_description(self, description):

TypeError: Sequence ID 42 must be of type `str`.
```
